src/supervisor/watchers/handlers/pagination.ts

```diff
@@ -0,0 +1,92 @@
+import { V1ObjectMeta } from '@kubernetes/client-node';
+import { IncomingMessage } from 'http';
+import * as sleep from 'sleep-promise';
+
+import { calculateSleepSeconds } from '../../kuberenetes-api-wrappers';
+import { V1NamespacedList } from './types';
+import type { IRequestError } from '../../types';
+
+const PAGE_SIZE = 100;
+
+export async function paginatedList<
+  T extends { metadata?: V1ObjectMeta },
+  TList extends { items: Array<T> },
+>(
+  namespace: string,
+  list: TList,
+  listPromise: V1NamespacedList<TList>,
+): Promise<{
+  response: IncomingMessage;
+  body: TList;
+}> {
+  let continueToken: string | undefined = undefined;
+
+  const pretty = undefined;
+  const allowWatchBookmarks = undefined;
+  const fieldSelector = undefined;
+  const labelSelector = undefined;
+
+  let incomingMessage: IncomingMessage | undefined = undefined;
+
+  loop: while (true) {
+    try {
+      const listCall = await listPromise(
+        namespace,
+        pretty,
+        allowWatchBookmarks,
+        continueToken,
+        fieldSelector,
+        labelSelector,
+        PAGE_SIZE,
+      );
+      incomingMessage = listCall.response;
+
+      if (Array.isArray(listCall.body.items)) {
+        const trimmedItems = listCall.body.items.map((item: T) => {
+          delete item.metadata?.managedFields;
+          return item;
+        });
+        list.items.push(trimmedItems);
+      }
+
+      continueToken = listCall.body.metadata?._continue;
+      if (!continueToken) {
+        break;
+      }
+    } catch (err) {
+      const error = err as IRequestError;
+
+      switch (error.code) {
+        case 'ECONNRESET':
+          const seconds = calculateSleepSeconds();
+          await sleep(seconds);
+          continue;
+        default:
+          break;
+      }
+
+      switch (error.response?.statusCode) {
+        case 410: // Gone
+          break loop;
+        case 429: // Too Many Requests
+        case 502: // Bad Gateway
+        case 503: // Service Unavailable
+        case 504: // Gateway Timeout
+          const seconds = calculateSleepSeconds(error.response);
+          await sleep(seconds);
+          continue;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  if (!incomingMessage) {
```

Comment:
can this be moved to just below L42?

Reply:
TS complains because there is a case where the first call returns 410 - we don't have an incoming message at that point